woocommerce/woocommerce-android
# [Coupons] Restrictions: Allowed Emails Input and Data Handling

Closes: #5540



### Description


This PR adds and allows editing of allowed emails in the "Allowed Emails" screen. 

This PR also adds proper data passing and handling:
- Passting allowed emails data from Coupon Restriction screen to Allowed Emails screen,
- Passing merchant-edited emails input from Allowed Emails screen to Coupon Restriction screen.

### Testing instructions

1. In wp-admin **Marketing > Coupons**, create a Coupon with email restriction. To add the restriction, create a Coupon and go to **Usage Restrictions > Allowed emails**.
2. Go to the app, make sure Coupons beta toggle is enabled.
3. Open the details for the Coupon from step 1.
4. Click on the Menu, then click on **"Edit coupon"**, then click **"Usage Restrictions"**.
5. Tap **"Allowed emails"** button.
6. Make sure the list of allowed emails added on step 1 is shown correctly on the field.
7. Enter an invalid email then tap the X button. Make sure error message is shown.
8. Enter a wildcard email (e.g: *@woocommerce.com) then tap the X button. Make sure the screen returns properly to Coupon Restrictions screen.
9. Enter some valid values, go back to **Coupon Restrictions** screen, then go back to **Edit Coupon** screen and tap **Save** button. Make sure it is saved properly and returns to **Coupon Details** screen.
10. On the Coupon Details screen, in the **Coupon Summary** section, make sure the newly saved email restrictions are shown with the **"Restricted to customers with emails"** label.

Optionally go back and play around with adding multiple values, emptying all values, and ensure the data is consistent between the two screens.

### Video

https://user-images.githubusercontent.com/266376/172356395-ecf61e25-3d04-48eb-ad3d-5a9a43acbce5.mov
